src/client.rs

```diff
@@ -297,6 +298,36 @@ pub extern "C" fn rustls_client_config_builder_load_roots_from_file(
     }
 }
 
+/// Set the ALPN protocol list to the given protocols. `protocols` must point
+/// to a buffer of `rustls_slice_bytes` (built by the caller) with `len`
+/// elements. Each element of the buffer must point to a slice of bytes that
+/// contains a single ALPN protocol from
+/// https://www.iana.org/assignments/tls-extensiontype-values/tls-extensiontype-values.xhtml#alpn-protocol-ids.
+///
+/// This function makes a copy of the data in `protocols` and does not retain
+/// any pointers, so the caller can free the pointed-to memory after calling.
+///
+/// https://docs.rs/rustls/0.19.0/rustls/struct.ClientConfig.html#method.set_protocols
+#[no_mangle]
+pub extern "C" fn rustls_client_config_builder_set_protocols(
+    builder: *mut rustls_client_config_builder,
+    protocols: *const rustls_slice_bytes,
+    len: size_t,
+) -> rustls_result {
+    ffi_panic_boundary! {
+        let config: &mut ClientConfig = try_mut_from_ptr!(builder);
+        let protocols: &[rustls_slice_bytes] = try_slice!(protocols, len);
+
+        let mut vv: Vec<Vec<u8>> = Vec::new();
+        for p in protocols {
+            let v: &[u8] = try_slice!(p.data, p.len);
+            vv.push(v.to_vec());
+        }
```

Comment:
Pending on the exact semantics of `try_slice!()`, maybe rewrite this as an iterator chain, that collects into `vv`? Would allow the code to allocate just once with the exact number of elements necessary. (Alternatively, use `Vec::with_capacity()` to allocate the right size.)

Reply:
I tried this but ran into some lifetime issues with [slice::from_raw_parts](https://doc.rust-lang.org/std/slice/fn.from_raw_parts.html). I believe it was inferring a too-short lifetime for resulting slice. I could possible annotate it at the cost of some complexity. `Vec::with_capacity` also sounds like a nice approach.

Reply:
I took another stab at this and realized it wasn't a lifetime issue. The problem is: when any of the `p.data` items is NULL, we want to early-return from the function with `NullParameter`. We can't do that from inside the closure of an iterator chain. We could make the closure return `Option<Vec<u8>>`, and then turn any Nones into an early return, but this seems like it would be more complicated.

Reply:
The standard solution for this is to return a `Result` type from the `map()` closure. `FromIterator` (the trait backing `collect()`) knows how to collect into a `Result`, so in this case you could collect into `Result<Vec<&[u8]>, NullParameter>` or something like that.

Reply:
Ah, thanks for the tip! The collect into Result is quite nice. Here's what I wound up with:

```rust
        let result = protocols.iter().map(|p| {
            if p.data.is_null() {
                Err(())
            } else {
                unsafe { Ok(slice::from_raw_parts(p.data, p.len).to_vec()) }
            }
        }).collect();
        match result {
            Ok(vv) => {
                config.set_protocols(&vv);
                rustls_result::Ok
            }
            Err(()) => rustls_result::NullParameter,
        }
```

14 lines, vs 7 for the procedural implementation (which I also find slightly more straightforward). If there's a way I can improve on that, lemme know!

Reply:
Ha, challenge accepted. Here's my shot (untested):

```rust
protocols.iter().map(|p| match p.data.is_null() {
    false => Ok(unsafe { slice::from_raw_parts(p.data, p.len).to_vec() }),
    true => Err(rustls_result::NullParameter),
}).collect().map(|vv| {
    config.set_protocols(&vv);
    rustls_result::Ok
}).unwrap_or_else(|e| e)
```

Also 7 lines, but I think I prefer the procedural version in this case.

Reply:
Ooh, nice code golf! That second `.map()` after collect is particularly clever.